src/cargo/sources/registry/mod.rs

```diff
@@ -497,9 +497,17 @@ impl<'cfg> RegistrySource<'cfg> {
             }
 
             // Once that's verified, unpack the entry as usual.
-            entry
-                .unpack_in(parent)
-                .chain_err(|| format!("failed to unpack entry at `{}`", entry_path.display()))?;
+            entry.unpack_in(parent).chain_err(|| {
```

Comment:
Would it be possible to structure this such that the error looks something roughly like this?

```
error: failed to download ...

Caused by:
  failed to unpack package ...

Caused by:
  failed to unpack entry at '{PATH}'

Caused by:
  '{PATH}' appears to contain a reserved windows path, it cannot be extracted on Windows

Caused by:
  ... More details from tar
```

So that the windows note is a separate item in the cause chain.  You'll need to use `let mut result = …` to be able to conditionally chain the error, then `result = result.chain_err(|| …);` inside the condition.

Reply:
Do you have an example of this kind of error chaining in the codebase?

Reply:
No, but I think it would look something like this:

```rust
let mut result = entry.unpack_in(parent).map_err(anyhow::Error::from);
if cfg!(windows) && restricted_names::is_windows_reserved_path(&entry_path) {
    result = result.chain_err(|| {
        format!(
            "`{}` appears to contain a reserved windows path, \
            it cannot be extracted on Windows",
            entry_path.display()
        )
    });
}
result.chain_err(|| format!("failed to unpack entry at `{}`", entry_path.display()))?;
```